### ISSUE TITLE
RegisterCodeFixesAsync never hits in custom CodeFixProvider

### ISSUE THREAD

**Issue by thomhurst** [dotnet/roslyn#68776]
I have used the Visual Studio Template to create an Analyzer + Code Fix. So everything should already be hooked up. I have the Analyzers, CodeFixes, Package, Test and Vsix projects.

On running my tests, I can debug through my analyzer, and I can see that my analyzer is reporting a diagnostic error.

However, the CodeFixProvider never hits the RegisterCodeFixesAsync. All it hits is the getter for the FixableDiagnosticIds, and that is matching the ID of what the analyzer is reporting.

My analyzer:

```csharp
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;

namespace ModularPipelines.Analyzers;

[DiagnosticAnalyzer(LanguageNames.CSharp)]
public class MissingDependsOnAttributeAnalyzer : DiagnosticAnalyzer
{
    public const string DiagnosticId = "MissingDependsOnAttribute";

    // You can change these strings in the Resources.resx file. If you do not want your analyzer to be localize-able, you can use regular strings for Title and MessageFormat.
    // See https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Localizing%20Analyzers.md for more on localization
    private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.AnalyzerTitle), Resources.ResourceManager, typeof(Resources));
    private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.AnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
    private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.AnalyzerDescription), Resources.ResourceManager, typeof(Resources));
    private const string Category = "Usage";

    private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description);

    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

    public override void Initialize(AnalysisContext context)
    {
        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
        context.EnableConcurrentExecution();
        
        context.RegisterSyntaxNodeAction(AnalyzeMissingDependsOnAttributes, SyntaxKind.InvocationExpression);
    }
        
    private void AnalyzeMissingDependsOnAttributes(SyntaxNodeAnalysisContext context)
    {
        if (context.Node is not InvocationExpressionSyntax invocationExpressionSyntax)
        {
            return;
        }

        if (invocationExpressionSyntax.Expression is not GenericNameSyntax genericNameSyntax)
        {
            return;
        }

        if (genericNameSyntax.Identifier.ValueText is not ("GetModule" or "WaitForModule"))
        {
            return;
        }

        var genericArgument = genericNameSyntax.TypeArgumentList.Arguments.First();

        var genericArgumentSymbol = context.SemanticModel.GetSymbolInfo(genericArgument).Symbol;
        
        if (genericArgumentSymbol is not INamedTypeSymbol namedTypeSymbol)
        {
            return;
        }

        var classContainingGetModuleCallSyntax = GetClassDeclarationSyntax(invocationExpressionSyntax);

        var classSymbol = context.SemanticModel.GetDeclaredSymbol(classContainingGetModuleCallSyntax);

        var attributes = classSymbol.GetAttributes();

        if (!attributes.Any(x => IsDependsOnAttributeFor(x, namedTypeSymbol)))
        {
            var properties = new Dictionary<string, string>
            {
                ["FullName"] = namedTypeSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat).Replace("global::", string.Empty)
            }.ToImmutableDictionary();

            context.ReportDiagnostic(Diagnostic.Create(Rule, context.Node.GetLocation(), properties, namedTypeSymbol.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat)));
        }
    }
        
    private ClassDeclarationSyntax GetClassDeclarationSyntax(InvocationExpressionSyntax invocationExpressionSyntax)
    {
        var parent = invocationExpressionSyntax.Parent;
            
        while (parent is not null)
        {
            parent = parent.Parent;
                
            if (parent is ClassDeclarationSyntax classDeclarationSyntax)
            {
                return classDeclarationSyntax;
            }
        }

        return null!;
    }

    private bool IsDependsOnAttributeFor(AttributeData attributeData, INamedTypeSymbol namedTypeSymbol)
    {
        var attributeClassName = attributeData.AttributeClass.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
        
        if (!attributeClassName.StartsWith("global::ModularPipelines.Attributes.DependsOnAttribute"))
        {
            return false;
        }

        return attributeData.ConstructorArguments.Any(x =>
            x.Type.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat) ==
            namedTypeSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat));
    }
}
```

My code fix provider:

```csharp
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Collections.Immutable;
using System.Composition;

namespace ModularPipelines.Analyzers;

[ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(MissingDependsOnAttributeCodeFixProvider)), Shared]
public class MissingDependsOnAttributeCodeFixProvider : CodeFixProvider
{
    public sealed override ImmutableArray<string> FixableDiagnosticIds
    {
        get { return ImmutableArray.Create(MissingDependsOnAttributeAnalyzer.DiagnosticId); }
    }

    public sealed override FixAllProvider GetFixAllProvider()
    {
        // See https://github.com/dotnet/roslyn/blob/main/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
        return WellKnownFixAllProviders.BatchFixer;
    }

    public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
    {
        var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
        
        var diagnostic = context.Diagnostics.First();
        var diagnosticSpan = diagnostic.Location.SourceSpan;

        // Find the type declaration identified by the diagnostic.
        var declaration = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<TypeDeclarationSyntax>().First();

        // Register a code action that will invoke the fix.
        context.RegisterCodeFix(
            CodeAction.Create(
                title: CodeFixResources.CodeFixTitle,
                createChangedDocument: c => AddAttribute(context, declaration, c),
                equivalenceKey: nameof(CodeFixResources.CodeFixTitle)),
            diagnostic);
    }

    private async Task<Document> AddAttribute(CodeFixContext context, TypeDeclarationSyntax typeDecl, CancellationToken cancellationToken)
    {
        var document = context.Document;
        var documentRoot = await document.GetSyntaxRootAsync(cancellationToken);

        var name = context.Diagnostics.First().Properties["FullName"]!;

        var attributes = typeDecl.AttributeLists.Add(
            SyntaxFactory.AttributeList(
                SyntaxFactory.SingletonSeparatedList(
                    SyntaxFactory.Attribute(
                            SyntaxFactory.IdentifierName("ModularPipelines.Attributes.DependsOnAttribute")
                        )
                        .WithArgumentList(
                            SyntaxFactory.AttributeArgumentList(
                                SyntaxFactory.SingletonSeparatedList(
                                    SyntaxFactory.AttributeArgument(
                                        SyntaxFactory.TypeOfExpression(SyntaxFactory.ParseTypeName(name))
                                    )
                                )
                            )
                        )
                )).NormalizeWhitespace());
            
        // TODO
        //documentRoot.DescendantNodes().OfType<UsingStatementSyntax>().Any(x => x.Statement.)
        
        return document.WithSyntaxRoot(
            documentRoot!.ReplaceNode(typeDecl, typeDecl.WithAttributeLists(attributes))
        );
    }
}
```

**Comment by kzu** [comment]
+100! Debugging CodeFixProvider seems (almost? entirely?) impossible. Productivity killer.

**Comment by CyrusNajmabadi** [comment]
Not able to repro a problem here. 

> +100! Debugging CodeFixProvider seems (almost? entirely?) impossible. 

it should be possible to attach a debugger normally.  Also, writing unit tests for a codefixprovider is very possible, and easily allows debugging into things.
